README.md
# Soulsborne

Soulsborne is a python text-based adventure game set in a World of Fantasy. 
The player's objective is to complete all 3 floors and escape the dungeon.


## Installation

Python 3.7 is needed to play this game. Click on [Python 3.7](https://www.python.org/downloads/) to be redirected to install python.


## Run the program

```python
python main.py to run main.py
python SimpleMenu.py to run simplemenu
```

## Contributions
Pull requests are welcome. 
Please do not hesitate to contact me for any major changes you would like to discuss about.

Please make sure to update tests as appropriate.

## License
[MIT] (https://choosealicense.com/licenses/mit/)


Changelog.md
# RPG: Soulsborne Changelog

vPrerelease
- Game is currently in a prerelease state.
- Added a README File
- Added a Changelog
- Added main.py file that lists current monsters, special monsters and weapons that are planned for the game (numerical list)
- Added Pseudocode.py that describes plans for creating the game
- Removed Pseudocode.py (Different repository)
- Updated main.py and README files
- Updated main.py file by adding Dictionaries for characters, locations and inventory
- Added a SimpleMenu
- Modified SimpleMenu to add quit option
   - Deleted Nested Dictionary to shorten code with only Loops
- Added a list of enemies to Soulsborne.
- Added a randomized map for each floor.

map.py
import random
from tabulate import tabulate


# Creating a list from elements in a dictionary
def append_list(dictionary, list):
    for x in dictionary:
        list.append(x)


# Choosing random tile to replace indices.
def random_tile(list, tile):
    x = random.choice(list)
    y = random.choice(x)
    a = list.index(x)
    b = x.index(y)
    list[a][b] = tile
    return(a, b)


# Making sure replaced tiles do not overlap.
def tile_replace(list, tile1, tile2, tile3, tile4):
    while random_tile(list, tile1) == random_tile(list, tile2):
        random_tile(list, tile1)
        random_tile(list, tile2)
        while random_tile(list, tile3) == random_tile(list, tile4):
                random_tile(list, tile3)
                random_tile(list, tile4)


# Randomly generate a 5x6 map with different tile types.
def generate_map(list):
    map = [[random.choice(list) for z in range(5)] for v in range(6)]
    tile_replace(map, "Boss", "Start", "Optional", "Safe Room")
    return map


# Printing the map with each floor.
def print_map(dictionary):
    for key in dictionary:
        map = dictionary[key]
        print(f"{key}")
        print(tabulate(map, tablefmt="grid"))
        print("\n")

# Floors in Soulsborne.
floors = {"Floor 1": "The beginner floor, peaceful with little to no harm.",
          "Floor 2": "Many adventureres have fallen here, be careful.",
          "Floor 3": "The end is nigh, nowhere is safe.",
          "Secret Floor": "Seekers receive Judgement, None Return."
          }

# Map tiles in Soulsborne.
map_tiles = {"Start": {
                        "description": "Starting area of the player.",
                        "abbrieviation": "S",
                        "action":
                        "Move to another location"},
             "Boss": {
                        "description": "Boss of the floor.",
                        "abbrievation": "BT",
                        "action":
                        "Defeat the boss to move to next floor."},
             "Enemy": {
                        "description": "Location of an Enemy.",
                        "abbrieviation": "ET",
                        "action":
                        "Defeat the enemy to continue."},
             "Items": {
                        "description": "Location of an item",
                        "abbrieviation": "IT",
                        "action":
                        "Pick up item to continue."},
             "Safe Room": {
                        "description": "Location of safe room",
                        "abbrieviation": "ST",
                        "action":
                        "Trade with merchant, rest or move"},
             " ": {
                        "description": "Blank location, nothing happens.",
                        "abbrieviation": "BT",
                        "action":
                        "Move to a different location"},
             "Optional": {
                            "description": "Optional Boss location",
                            "abbrieviation": "OP",
                            "action":
                            "Defeat boss to continue."}}


# List of floors
floor_level = []
append_list(floors, floor_level)
# List of tiles and Removing start, boss, optional and safe floors.
tile_types = []
append_list(map_tiles, tile_types)
tile_types.remove("Boss")
tile_types.remove("Start")
tile_types.remove("Optional")
tile_types.remove("Safe Room")

# Using loop function to help print floors into map.
main_map = {}
for floor in floor_level:
    floors = generate_map(tile_types)
    main_map[floor] = floors

# Printing the Main Map.
(print_map(main_map))


main.py
# Course: CS 30
# Period: 1
# Date Created: 20/09/21
# Date Modified: 20/09/29
# Name: Michael Nguyen
# Description: Soulsborne dictionaries of characters, locations and inventory.

# Nested dictionary of characters in Soulsborne.
characters = {
    'hero': {
      'who': 'The protagonist of the adventure. (You!)',
      'description': 'is an ordinary hero.',
      'health points': 30,
    },
    'merchant': {
      'who': 'An NPC that sells you items.',
      'description': 'is just a merchant, nothing special.',
      'health points': 0,
    },
    'floor 1 boss': {
      'who': 'The boss of floor 1.',
      'description': 'is a weak introductory boss.',
      'health points': 30,
    },
    'floor 2 boss': {
      'who': 'The boss of floor 2.',
      'description': 'is a huge skeleton boss.',
      'health points': 45,
    },
    'floor 3 boss': {
      'who': 'The final boss, gatekeeper of floor 3.',
      'description': 'is the final boss.',
      'health points': 65,
    },
    'secret floor boss': {
      'who': 'An optional secret boss.',
      'description': 'is an optional secret boss.',
      'health points': 99,
    },
    'skeleton': {
      'who': 'Common mob found on floor 1.',
      'description': 'is a commonly found mob.',
      'health points': 15,
    },
    'slime': {
      'who': 'Common mob found on floor 1.',
      'description': 'is a weak starting enemy.',
      'health points': 5,
    },
    'hollow knight': {
      'who': 'Secret enemy found on floor 1.',
      'description': 'is an optional boss.',
      'health points': 25,
    },
}

# Looping Soulsborne characters's traits.
for characters, character_info in characters.items():
    who = (character_info['who'])
    print(f"\n{characters.title()}: {who}")
    description = (character_info['description'])
    print(f"  The {characters.title()} {description}")
    health_points = (character_info['health points'])
    print(f"  The {characters.title()} has {health_points} health points.")


print()
print()

# Dictionary of floors.
floors = {
    'floor 1': 'A relatively peaceful floor for beginners.',
    'floor 2': 'Floor with dangerous monsters, proceed with caution. ',
    'floor 3': 'The end is nigh, you sense danger all around.',
    'secret floor': 'None have survived Judgement.',
}

# Looping Soulsborne floors.
for floor, space in floors.items():
    print(f"{floor.title()}: {floors[floor]}")


print()
print()

# Dictionary of different tiles.
tiles = {
    'enemy tile': 'Location of the enemy. Defeat the enemy to proceed.',
    'boss tile': 'Location of the boss. Defeat boss to proceed to next floor.',
    'blank tile': 'Just an empty tile. Move to another location.',
    'safe tile': 'Location of safe room. Rest or move to another location.',
    'item tile': 'Location of an item. Could be a weapon or healing.',
}

# Looping different tiles in Soulsborne.
for tile, space in tiles.items():
    print(f"You are on a {tile.title()}: {tiles[tile]}")

print()
print()

# Nested dictionary of the inventory.
inventory = {
    'long sword': {
        'damage': 10,
        'description': 'A standard sword.',
        'protection': 0,
    },
    'basic shield': {
        'damage': 0,
        'description': 'Shield for blocking.',
        'protection': 10,
    },
    'fire spell': {
        'damage': 10,
        'description': 'Standard magic spell.',
        'protection': 0,
    },
    'gold coins': {
        'damage': 0,
        'description': 'Currency for bartering.',
        'protection': 0,
    }
}

# Looping each item in player's inventory.
print("Hero's Inventory:")
for inventory, inventory_info in inventory.items():
    print(f"\n* {inventory.title()}:")
    description = (inventory_info['description'])
    print(f"    Description: {description}")
    damage = (inventory_info['damage'])
    print(f"    Damage: {damage}")
    protection = (inventory_info['protection'])
    print(f"    Protection: {protection}")


enemies.py
class Enemy():
    def __init__(self):
        raise NotImplementedError("Do not create raw Enemy objects")

    def __str__(self):
        return self.name

    def is_alive(self):
        return self.hp > 0

# Floor 1 Enemies


# Slime enemy detailing its hp and damage
class Slime(Enemy):
    def __init__(self):
        self.name = "Slime"
        self.hp = 10
        self.damage = 5


# Goblin enemy detailing its hp and damage
class Goblin(Enemy):
    def __init__(self):
        self.name = "Goblin"
        self.hp = 15
        self.damage = 10


# Lizard enemy detialing its hp and damage
class Lizard(Enemy):
    def __init__(self):
        self.name = "Lizard"
        self.hp = 15
        self.damage = 5


# Hollow Knight optional boss detailing its hp and damage
class Hollow_Knight(Enemy):
    def __init__(self):
        self.name = "Hollow Knight"
        self.description = "The Wandering Knight who since lost his name..."
        self.hp = 25
        self.damage = 15


# King Slime boss detailing its hp and damage
class King_Slime(Enemy):
    def __init__(self):
        self.name = "King Slime"
        self.description = "This is the King of all Slimes!"
        self.hp = 20
        self.damage = 10

# Floor 2 Enemies


# Skeleton enemy detailing its hp and damage
class Skeleton(Enemy):
    def __init__(self):
        self.name = "Skeleton"
        self.hp = 20
        self.damage = 15


# Undead enemy detailing its hp and damage
class Undead(Enemy):
    def __init__(self):
        self.name = "Undead Zombie"
        self.hp = 5
        self.damage = 30


# Troll enemy detailing its hp and damage
class Troll(Enemy):
    def __init__(self):
        self.name = "Troll"
        self.hp = 25
        self.damage = 15


# Berserkerlot optional boss detailing its hp and damage 
class Beserkerlot(Enemy):
    def __init__(self):
        self.name = "Pinwheel"
        self.description = "A Knight that has lost all worldly desires..."
        self.hp = 45
        self.damage = 25


# Arturia boss detailing its hp and dmdage
class Arturia(Enemy):
    def __init__(self):
        self.name = "King Arturia"
        self.description = "The King is no exception to Absurdity."
        self.hp = 65
        self.damage = 25

# Floor 3 Enemies


# Hunter enemy detailing its hp and damage
class Hunter(Enemy):
    def __init__(self):
        self.name = "Hunter"
        self.hp = 35
        self.damage = 20


# Skirmisher enemy detailing its hp and damage
class Skirmisher(Enemy):
    def __init__(self):
        self.name = "Skirmisher"
        self.hp = 40
        self.damage = 25


# Great Wolf enemy detailing its hp and damage
class Great_Wolf(Enemy):
    def __init__(self):
        self.name = "Great Wolf"
        self.hp = 30
        self.damage = 30


# Gwyn optional boss detailing its hp and damage
class Gwyn_Upholder_of_Peace(Enemy):
    def __init__(self):
        self.name = "Gwyn, Upholder of Peace"
        self.description = "'You dare to disturb the everlasting peace?!'"
        self.hp = 55
        self.damage = 30


# Soul of Cinder "final" boss detailing its hp and damage
class Soul_of_Cinder(Enemy):
    def __init__(self):
        self.name = "Soul of Cinder"
        self.description = "The End of your Journey is nigh."
        self.hp = 70
        self.damage = 35

# Secret Floor Enemies

# Vessel enemy detailing its hp and damage.
class Vessel(Enemy):
    def __init__(self):
        self.name = "Vessel"
        self.description = "The Souls of the condemned served as vessels."
        self.hp = 5
        self.damage = 0


# King Manus optional secret boss detailing its hp and damage
class King_Manus_The_First_Hero(Enemy):
    def __init__(self):
        self.name = "King Manus, The First Hero"
        self.description = "None have survived Judgement."
        self.hp = 99
        self.damage = 50


# The Vessel secret final boss detailing its hp and damage
class The_Holy_Grail(Enemy):
    def __init__(self):
        self.name = "The Holy Grail"
        self.description = "The Holy Grail upholding the entire world."
        self.hp = 1
        self.damage = 0


Pseudocode.py
# Pseudocode Soulsborne:
# player character spawns in
# task the player the goal/objective that the player must complete
# give player character basic weapons, armor and consumables
# each floor contains boss room and safe room
# players can roam around fighting monsters and try to find those rooms
# safe rooms provide a merchant and safe area to rest at
# boss rooms contain a boss that must be defeated to continue on to the next floor


# Pseudocode player character:
# player character is alive
# hp values remain at 30
# if player gets hit, remove hp values (hp value varies per enemy)
# if player dies, game ends
# if player attacks, generate random number generator for attack, miss or critical hti


# Pseudocode monster (applies to special monster):
# monster is alive
# hp values remain at full (hp value varies per monster)
# if monster gets hit, remove hp values based on player character's attack
# if monster dies, drop loot and exp
# if monster attacks, generate random number generator for attack, miss or critical hit


# Pseudocode floor boss:
# Floor boss is alive
# hp values remain at full (hp value varies per boss)
# if boss gets hit, remove hp values based on character's attack
# if boss dies, grant access to proceeding floor
# if final boss dies, game ends
# if boss attacks, generate random number generator for attack, miss or critical hit


# Pseudocode inventory:
# player character receives item
# type in text box to access inventory
# add sections for armor, weapons and consumables
# players can access different inventories by typing in different number values


# Pseudocode floor:
# generate floor layout
# provide a map of the floor in safe room
# safe room will always be close to player character spawn
# randomly generate a chance for monster to spawn when player character enters room
# generate random monster (based on %) encounter 
# one room on the floor  will 100% always have a special monster


# Pseudocode items:
# if item is dropped from monster, add to inventory
# consumable items will be consumed from the inventory
# key items will be used, but still retain inside the inventory
# add a healing value to items
# when item is consumed add to player character hp value


# Pseudocode weapons:
# depending on weapon, it will have a base damage value
# weapons can be sold
# weapons cannot be consumed
# if enemy is hit by a weapon with a damage value of 3, enemy hp value decreases by 3


# Pseudocode armor: 
# armor raises base hp value of player character
# does not regen hp values
# armor can be sold
# armor cannot be consumed
# removing armor resets hp value to base of 30


# Pseudocode ending:
# display ending based on in game story progression
# 2 endings and one secret ending
# after beating the game, display win screen
# run main.py to restart the game

playerinventory.py
inventory = {"Hero": { 'long sword': 
        'damage': 10,
        'description': 'A standard sword.',
        'protection': 0,
    },
    'basic shield': {
        'damage': 0,
        'description': 'Shield for blocking.',
        'protection': 10,
    },
    'fire spell': {
        'damage': 10,
        'description': 'Standard magic spell.',
        'protection': 0,
    },
    'gold coins': {
        'damage': 0,
        'description': 'Currency for bartering.',
        'protection': 0,
    }
}
def player_inventory(player, inventory):
    protection_items = []
    weapons = []
    for item in inventory[player]:
        description = inventory[player][item][description]
        damage = inventory[player][item][damage]
        protection = inventory[player][item][protection]
        print(f" {player}'s {item} - {description}")
        print(f" {damage}: {damage}")
        print(f" {protection}: {protection}")
        if protection != 0 and damage == 0:
            protection_items.append(item)
        elif damage != 0:
            weapons.append(item)
        return(protection_items, weapons)

print(inventory)


SimpleMenu.py
# Course: CS 30
# Period: 1
# Date Created: 20/09/21
# Date Modified: 20/10/06
# Name: Michael Nguyen
# Description: SimpleMenu updated for Continuous Gameplay.



print("Valid actions for current location:")
print("Type in 'quit' to quit out of the game.")

# Looping and printing list of directions.
print("Which direction would you like to move in?: ")
directions = ['* North', '* South', '* East', '* West']
for pole in directions:
    print(pole)

# Looping and printing list of actions.
print("\nYour move, what is your action?: ")
actions = ['* Attack', '* Defend', '* Heal', '* Wander', '* Quit']
for action in actions:
    print(action)

# Various actions the player can take. Actions are looped.
# Added an option to quit the game using an active flag.
active = True
while active:
    prompt = input("\nAction: ").lower().strip()

    if prompt == 'quit':
        active = False

    if prompt == 'north':
        print("You move North.")
    elif prompt == 'south':
        print("You move South.")
    elif prompt == 'east':
        print("You move East.")
    elif prompt == 'west':
        print("You move West.")
    elif prompt == 'attack':
        print("You attack the enemy.")
    elif prompt == 'defend':
        print("You stand your ground and defend.")
    elif prompt == 'heal':
        print("You heal the damage taken.")
    elif prompt == 'wander':
        print("You wander and explore around.")
    elif prompt == 'quit':
        print("You have quit the game.")
    else:
        print("That is an invalid action!")
